### PR TITLE
Add connectivity transforms

### PR DESCRIPTION
**Pull request by martin-carrasco**
Adds the connectivity transforms for:

- Incidences
- Adjacencies
- Up Laplacian
- Down Laplacian
- Co-adjacencies

### CODE REVIEW

File: mantra/representations/simplicial_connectivity.py [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+        assert "triangulation" in data
+        simplex_trie = SimplexTrie()
+        for t in data.triangulation:
+            t_ = frozenset(t)
+            simplex_trie.insert(sorted(t_))
+        data.simplex_trie = simplex_trie
+        return data
+
+
+class ConnectivitySCTransform(T.BaseTransform):
```
Comment (Pseudomanifold) [inline mantra/representations/simplicial_connectivity.py]: Can we find a better name for this? I would first suggest to write out "SC", i.e. "SimplicialComplex." I would also suggest to remove the `Transform` suffix, unless we need it. Since we have to import these transforms/representations, my understanding is that they only ever appear in a couple of places anyway, and I don't think the suffix is required.

This is **not** a good justification, but `pytorch-geometric` does not use this suffix very often (see [documentation](https://pytorch-geometric.readthedocs.io/en/stable/modules/transforms.html))

File: mantra/representations/simplicial_connectivity.py [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+                    data[connectivity_name] = torch.zeros(
+                        [shape[rank_idx], shape[rank_idx]],
+                        layout=torch.sparse_coo,
+                    ).coalesce()
+
+        return data
+
+
+class IncidenceSCTransform(ConnectivitySCTransform):
+    """Add incidences of a simplicial complex."""
```
Comment (Pseudomanifold) [inline mantra/representations/simplicial_connectivity.py]: Good to leave it "as-is" now, but could you later on provide a brief example?

File: mantra/representations/simplicial_connectivity.py [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+            )
+        if not self.signed:
+            L_up = abs(L_up)
+        if self.index:
+            return row, L_up
+
+        return L_up
+
+
+class DownLaplacianSCTransform(ConnectivitySCTransform):
```
Comment (Pseudomanifold) [inline mantra/representations/simplicial_connectivity.py]: I believe this would also benefit from being more "verbose" as a name (minus the "Transform" part).

File: mantra/representations/simplicial_connectivity.py [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+            The datastructure containing a simplex.
+        rank: int
+            Rank r of the skeleton.
+        Returns:
+            List[Simplex]
+
+        """
+        return sorted(node.simplex for node in simplex_trie.skeleton(rank))
+
+    def generate_matrix(
```
Comment (Pseudomanifold) [inline mantra/representations/simplicial_connectivity.py]: Let's make this abstract to ensure that it's clear that this needs implementation.

File: mantra/representations/simplex.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+from typing_extensions import Self
+
+
+__all__ = ["Simplex"]
+
+ElementType = TypeVar("ElementType", bound=Hashable)
+
+
+@total_ordering
+class Simplex(Generic[ElementType]):
```
[the remark this answers is not in the document]
Comment (martin-carrasco) [reply mantra/representations/simplex.py]: With the little time it took that would definitevly be a feat! I took the code from TopoNetX, also added a reference now :)

File: mantra/representations/simplex.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+
+        Returns
+        -------
+        str
+            A string representation of the simplex.
+        """
+        return f"Simplex({tuple(self.elements)})"
+
+    def __str__(self) -> str:
+        """Return string representation of the simplex.
```
[the remark this answers is not in the document]
Comment (martin-carrasco) [reply mantra/representations/simplex.py]: Done!

File: mantra/representations/simplex.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+        Returns
+        -------
+        str
+            A string representation of the simplex.
+        """
+        return f"Nodes set: {tuple(self.elements)}"
```
[the remark this answers is not in the document]
Comment (martin-carrasco) [reply mantra/representations/simplex.py]: I changed the str to `"Nodes:....". Is that what you meant ?

File: mantra/representations/__init__.py [hunk 8]
```diff
@@ -1,4 +1,13 @@
...
+    CoadjacencySCTransform,
+)
 
-__all__ = ["DualGraph", "IncidenceSCTransform", "AdjacencySCTransform", "CoadjacencySCTransform"]
+__all__ = [
+    "DualGraph",
+    "IncidenceSCTransform",
+    "AdjacencySCTransform",
+    "CoadjacencySCTransform",
+]
```
[the remark this answers is not in the document]
Comment (martin-carrasco) [reply mantra/representations/__init__.py]: Done!

File: mantra/representations/simplicial_connectivity.py [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+
+
+class AddSimplexTrie(T.BaseTransform):
+    """Add simplex trie to object.
+
+    Takes in a `Data` object and computes a `SimplexTrie` for it.
+    Then it adds it to it's storage to save up on computation
+    of connectivity matrices.
+
+
```
[the remark this answers is not in the document]
Comment (martin-carrasco) [reply mantra/representations/simplicial_connectivity.py]: Done!

File: mantra/representations/simplicial_connectivity.py [hunk 10]
```diff
@@ -0,0 +1,300 @@
+from typing import Dict, List
+from networkx import incidence_matrix
+import torch
+import numpy as np
+from scipy.sparse import csr_matrix
+import scipy
+
+import torch_geometric.transforms as T
```
[the remark this answers is not in the document]
Comment (martin-carrasco) [reply mantra/representations/simplicial_connectivity.py]: Done!

File: mantra/representations/simplicial_connectivity.py [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+from torch_geometric.data import Data
+
+from mantra.representations.simplex_trie import SimplexTrie, Simplex
```
[the remark this answers is not in the document]
Comment (martin-carrasco) [reply mantra/representations/simplicial_connectivity.py]: Done!

File: mantra/representations/simplex_trie.py [hunk 12]
```diff
@@ -0,0 +1,476 @@
...
+
+        # update label lists
+        if node.depth not in self.label_lists:
+            self.label_lists[node.depth] = {}
+        if label in self.label_lists[node.depth]:
+            self.label_lists[node.depth][label].append(node)
+        else:
+            self.label_lists[node.depth][label] = [node]
+
+        # update shape property
```
[the remark this answers is not in the document]
Comment (martin-carrasco) [reply mantra/representations/simplex_trie.py]: So the idea behind this is the follow, badly named but anyways...

`label_list` contains lists index by the rank of the simplices. For a rank `r`, `label_list[r]` contains a list of `r` simplices.

In line 342:
   If there is already a `r+1` simplex that has a particular `r` simplex in it's label list, then when this `r` simplex is added as part of ANOTHER `r+1` simplex it appends the to the label the simplex object it corresponds too.
 In 345:
  This is the case when the first `r` simplex of this ficticious `r+1` has appeared so we set the label to point to only 1.
  
  Concretely if I have previously added a triangle `(0,1,2)`, then it will add `(0,1), (1,2), (0,2)` as simplices. Then `label_lists[1][(0,1) = SimplexNode((0,1), (0,1,2)[AS A SIMPLEX NODE OBJECTS]')`
  
  
   If then I make another adition of a triangle `(0,1,3)` then `label_list[1][0,1)]` gets another addition of ANOTHER SimplexNode which has a parent in `(0,1,3)` .
   
   Hope this clarifies it, I can also make this into a docstring example

File: mantra/representations/simplicial_connectivity.py [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+            )
+        if not self.signed:
+            L_down = abs(L_down)
+        if self.index:
+            return column, L_down
+        return L_down
+
+
+class AdjacencySCTransform(ConnectivitySCTransform):
+    """Add adjacencies of a simplicial complex."""
```
[the remark this answers is not in the document]
Comment (martin-carrasco) [reply mantra/representations/simplicial_connectivity.py]: Yes!

File: mantra/representations/simplicial_connectivity.py (hunk 2 above)
Comment (martin-carrasco) [reply mantra/representations/simplicial_connectivity.py]: Yes!

File: mantra/representations/simplicial_connectivity.py [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+                if "incidence" in self.connectivity_name:
+                    data[connectivity_name] = torch.zeros(
+                        [shape[idx_low_simp], shape[rank_idx]],
+                        layout=torch.sparse_coo,
+                    ).coalesce()
+                elif "adjacency" in self.connectivity_name:
+                    data[connectivity_name] = torch.zeros(
+                        [shape[rank_idx], shape[rank_idx]],
+                        layout=torch.sparse_coo,
+                    ).coalesce()
```
[the remark this answers is not in the document]
Comment (Pseudomanifold) [reply mantra/representations/simplicial_connectivity.py]: OK, thanks!

File: mantra/representations/simplex.py [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+        other : Any
+            The other simplex to compare with.
+
+        Returns
+        -------
+        bool
+            True if this simplex comes before the other simplex in the lexicographic order.
+        """
+        if not isinstance(other, Simplex):
+            return NotImplemented
```
[the remark this answers is not in the document]
Comment (martin-carrasco) [reply mantra/representations/simplex.py]: Done!
